lib/spack/spack/test/sbang.py

```diff
@@ -191,17 +191,20 @@ def test_shebang_handles_non_writable_files(script_dir, sbang_line):
 
 def check_sbang_installation():
     sbang_path = sbang.sbang_install_path()
-    sbang_bin_dir = os.path.dirname(sbang_path)
     assert sbang_path.startswith(spack.store.store.unpadded_root)
 
     assert os.path.exists(sbang_path)
     assert fs.is_exe(sbang_path)
 
-    status = os.stat(sbang_path)
-    assert (status.st_mode & 0o777) == 0o755
 
+def check_sbang_bin_dir_permissions():
```

Comment:
Why not check the directory permissions at ~line 219? 

Also, you could keep the directory perms check with `check_sbang_installation` by restoring line 194, grabbing the expected mode for the directory using `config_mode = spack.package_prefs.get_package_dir_permissions(spack.spec.Spec("all"))` and using a single assert `assert (status.st_mode & 0o777) == config_mode` for the directory.

Reply:
I am getting config_mode = 0o700 when I do it this way. Not sure why.

Reply:
That's causing it to fail the tests.

Reply:
Those pesky `umasks`.  And there isn't a good/safe way to check the current value without changing it, which can be problematic.

Reply:
Do the checks you removed at lines 200-201 fail for you?

Reply:
Manually checking the sbang changes as of `bb9299d` works for me but I experienced the same issues with tests.

I have a fix for the tests but don't have permissions to push them to your branch.  

Basically I am suggesting replacing the current contents of lines 192 on of `lib/spack/spack/test/sbang.py` with the code below.  The fixture should ensure consistent permissions.  (I'm not specifying a group here because we can't guarantee a benign group across all systems.)  I'm curious of this will work for you *and* if it will pass the PR checks.

```
@pytest.fixture(scope='function')
def configure_world_perms():
    conf = syaml.load_config("""\
all:
  permissions:
    read: world
    write: world
""")
    spack.config.set('packages', conf, scope='user')

    yield


def check_sbang_installation():
    sbang_path = sbang.sbang_install_path()
    sbang_bin_dir = os.path.dirname(sbang_path)
    assert sbang_path.startswith(spack.store.store.unpadded_root)

    assert os.path.exists(sbang_path)
    assert fs.is_exe(sbang_path)

    status = os.stat(sbang_bin_dir)
    mode = (status.st_mode & 0o777)
    assert mode == 0o755, 'Unexpected {0}'.format(oct(mode))

    status = os.stat(sbang_path)
    mode = (status.st_mode & 0o777)
    assert mode == 0o777, 'Unexpected {0}'.format(oct(mode))


def test_install_sbang(install_mockery, configure_world_perms):
    sbang_path = sbang.sbang_install_path()
    sbang_bin_dir = os.path.dirname(sbang_path)

    assert sbang_path.startswith(spack.store.store.unpadded_root)
    assert not os.path.exists(sbang_bin_dir)

    sbang.install_sbang()
    check_sbang_installation()

    # put an invalid file in for sbang
    fs.mkdirp(sbang_bin_dir)
    with open(sbang_path, "w") as f:
        f.write("foo")

    sbang.install_sbang()
    check_sbang_installation()

    # install again and make sure sbang is still fine
    sbang.install_sbang()
    check_sbang_installation()
```

Reply:
The following import is also needed for the above to work:

```
import spack.util.spack_yaml as syaml
```

Reply:
You're right, that did fix the tests (and I added your changes in 527338b). However, that only ever tests a situation where the group permissions are not set. Should we add another test case to verify the installation of `sbang` with group write permissions?

Reply:
Configuring to 'world', IIRC, automatically sets group permissions.

AFAIK The only way to make such a test portable would be to monkey patch the system call(s) for a test group.  I'm not sure it's worth the extra effort.  But it's up to you.

Reply:
Not too worried about it, this seems good to me.